### PR TITLE
Sketcher/Toponaming: import SketchObject::checkSubName() from LS3

### LINKED ISSUE #16422: [Regression] Sketcher: No visual feedback when external line is selected (v1.0 RC1)
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Problem description

In the sketcher workbench, when using external reference lines (not only straight lines, anything), the lines are properly highlighted in the preselection mode color, but do not change color when selected. External points are properly changing color when selected.

### Full version info

```shell
OS: Gentoo Linux
Word size of FreeCAD: 64-bit
Version: 1.0.0RC1.38642 (Git) AppImage
Build type: Release
Branch: (HEAD detached at 1.0rc1)
Hash: 60a251354e203bdb924e2190fb9f5e18b48d7362
Python 3.11.9, Qt 5.15.13, Coin 4.0.2, Vtk 9.2.6, OCC 7.7.2
Locale: English/United Kingdom (en_GB)
Installed mods: 
  * freecad.gears 1.2.0
  * persistenttoolbars
```

### Subproject(s) affected?

Sketcher

### Anything else?

Preselecting the line (configured preselection color is light blue here)
Selection_003

Selected a point from the external geometry
Selection_004

Selected the same vertical line as in the first image (believe me : the line IS selected)
Selection_005

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by CalligaroV**
related to https://github.com/FreeCAD/FreeCAD/issues/13177#issuecomment-2307792664

the imported code restores the synchronization of the selection of external edges between the 3D view and the Task Panel

as for https://github.com/FreeCAD/FreeCAD/issues/16422#issuecomment-2344255862, this PR fixes #16422

### CODE REVIEW

**Comment by CalligaroV** [comment]
@bgbsww @chennes @Pesc0, I see that the code imported for this PR were deleted in https://github.com/FreeCAD/FreeCAD/pull/15194/commits/0e5690d732521384c081f64cae90f381bfbf7eba

are there any reasons to not import it?
without that this PR is useless but it looks needed to fix the selection/highlighting issue mentioned in https://github.com/FreeCAD/FreeCAD/issues/13177#issuecomment-2307792664

TIA!

**Comment by bgbsww** [comment]
@CalligaroV IIRC that was part of a code cleanup to make what was going on clearer, and there was nothing at that point using that code.  So the only reason is about clarity to somebody new reading it, and not functionality.

**Comment by CalligaroV** [comment]
Many thanks @bgbsww for the reply.

So can we keep this PR open (obviously also if it fixes the related issue)?

File: src/Mod/Sketcher/App/SketchObject.cpp [hunk 1]
```diff
@@ -11350,26 +11350,74 @@ Data::IndexedName SketchObject::checkSubName(const char *subname) const{
...
+        }
+    }
+    if (geo && GeometryFacade::getId(geo) == id) {
+        char sep;
+        int posId = static_cast<int>(PointPos::none);
+        if ((iss >> sep >> posId) && sep == 'v') {
+            int idx = getVertexIndexGeoPos(geoId, static_cast<PointPos>(posId));
+            if (idx < 0) {
+                FC_ERR("invalid subname " << subname);
+                // return sub;
```
Comment (bgbsww) [inline src/Mod/Sketcher/App/SketchObject.cpp:11408]: Remove, maybe change return.

**Review by bgbsww — COMMENTED** [review]
Minor stuff.  LGTM, just pending confirmation that it fixes bug.

File: src/Mod/Sketcher/App/SketchObject.cpp [hunk 2]
```diff
@@ -11350,26 +11350,74 @@ Data::IndexedName SketchObject::checkSubName(const char *subname) const{
...
-    default: {
-        const char *dot = strchr(mappedSubname,'.');
-        if(dot)
-            mappedSubname = dot+1;
-        return Data::IndexedName(mappedSubname, types, false);
-    }}
-
-    // TNP July '24: omitted code related to external and internal sketcher stuff implemented in LS3
+        // return sub;
+        return Data::IndexedName();
```
[the remark this answers is not in the document]
Comment (CalligaroV) [reply src/Mod/Sketcher/App/SketchObject.cpp:11378]: I tried now to apply your hint.

Still looks working, the only strange thing I noticed is that during compilation I get the following warning (twice, both for line 11378 and 11409):
```
warning: converting to ‘Data::IndexedName’ from initializer list would use explicit constructor ‘Data::IndexedName::IndexedName(const char*, int)'
```

If that's not an issue I can push another commit with your hints applied

**Comment by bgbsww** [comment]
Huh.  Then I'd leave it alone, just get rid of the commented out lines.  I
was trying to make the compiler / linter happier.

**Comment by Syres916** [comment]
I have been testing this PR, so far just one glitch found, I took the example file _PartDesignExample_ and modified it by creating a new Sketch on the face on Body.Pocket002.Face6. I created a rectangle which I firstly converted to Construction, added a couple of External Geometry lines and later reverted the rectangle back to 'real' geometry. I constrained the top edge of the rectangle using DistanceX without issue but on selecting the vertical left hand Edge1 in order to apply a DistanceY constraint I received these errors in the Report View:

```
17:55:16  <Sketch> SketchObject.cpp(11419): cannot find subname ;g10;SKT.Edge1
17:55:16  <Sketch> SketchObject.cpp(11419): cannot find subname ;g10;SKT.Edge1
17:55:16  <Sketch> SketchObject.cpp(11419): cannot find subname ;g10;SKT.
17:55:16  <Sketch> SketchObject.cpp(11419): cannot find subname ;g10;SKT.
17:55:16  <Sketch> SketchObject.cpp(11419): cannot find subname ;g10;SKT.Edge1
17:55:16  <Sketch> SketchObject.cpp(11419): cannot find subname ;g10;SKT.Edge1
```

the selection in the Python Console all looks correct:

`# Gui.Selection.addSelection('PartDesignExample','Body','Sketch004.Edge1',-45.1748,-20.008,-75.8102,False)`

Of course, saving and then re-opening the file everything is fine but does there need to be an extra Recompute/rebuild of the element map of the Sketch after the geometry has been created?

Screenshot from 2024-09-11 18-05-28

Screenshot from 2024-09-11 18-04-59

**Comment by bgbsww** [comment]
Thanks, @Syres916 .  If you just force a recompute on the sketch does that also clean up the problem?

Two possible paths in my head: determine that that error is unnecessary and we should just put up a warning and proceed; or look into forcing recomputes after sketch geom changing events like toggling const geo.

**Comment by Syres916** [comment]
> If you just force a recompute on the sketch does that also clean up the problem?

Yep

In recreating the errors from scratch I realised one crucial step that I'd forgot, I had deleted an isolated _Line_ segment and I don't believe it makes any difference where in the sequence of operations after all the geometry has been created that the deletion occurs it will still generate the errors if the _Recompute_ hasn't been actioned after the delete.


> look into forcing recomputes after sketch geom changing events like toggling const geo.

this seems the way to go certainly with deletions

**Comment by CalligaroV** [comment]
Thanks as always @bgbsww @Syres916 for your contribution!

If I understand correctly, the solution is to modify the code related to the deletion of some geometries.
I can add also this modification but it may take me some time to figure out the best way to do that.
@bgbsww do you have some hints about the element-related method(s) I should call?

Side note: even if this PR is causing the error/warning messages, the root cause of the issue looks related to another part of the code.
@chennes, should I apply the modification that fixes it in this PR or open another one dedicated for that?

**Comment by CalligaroV** [comment]
as for https://github.com/FreeCAD/FreeCAD/issues/16422#issuecomment-2344255862, this PR fixes #16422

**Comment by chennes** [comment]
I don't have a strong opinion on the matter -- if the three of you agree that this PR represents forward progress (albeit incomplete) I am happy to review and merge. If you feel like it really needs the next modification before going in, I will hold off.

**Comment by bgbsww** [comment]
SketchObject::delGeometry - the end of the method and the two methods after it are where I would start poking.  I'd also see if executing a solve corrected things.  I'm mostly suspicious that the geo coming into buildShape has not been updated correctly, so back tracking from there might help.

I don't see any explicit elementMap removals, which makes me wonder if it's only ever resetElementMaps followed by adding back in the geo that happens.

**Comment by CalligaroV** [comment]
Thanks for the feedback @chennes @bgbsww!

I'll keep this PR as is and open a new one for the modifications needed to solve the issue @Syres916 found.
If anything else comes up because of this PR just let me know.

Thanks!

**Pull request merged**